datafusion/src/optimizer/constant_folding.rs

```diff
@@ -101,7 +105,30 @@ impl OptimizerRule for ConstantFolding {
                             // fold constants and then simplify
                             .rewrite(&mut const_evaluator)?
                             .rewrite(&mut simplifier)?;
-                        Ok(new_e)
+
+                        let new_name = &new_e.name(plan.schema());
+
+                        // Some plans will be candidates in projection pushdown rule to
+                        // trim expressions based on expression names. We need to keep
+                        // expression name for them.
+                        let is_plan_for_projection_pushdown = matches!(
+                            plan,
+                            LogicalPlan::Window { .. }
```

Comment:
Why only those? 

What about `SELECT 1+1`.

Currently this outputs:

```
❯ SELECT 1+1
;
+----------+
| Int64(2) |
+----------+
| 2        |
+----------+
1 row in set. Query took 0.001 seconds.
```

I would assume it should keep the `Int64(1) + Int64(1)` here instead.

Reply:
For example, for Project, it will create many (looks redundant) aliases. Some looks okay but some looks really weird, e.g. some failed tests:

```
Projection: #test.a, #test.d, NOT #test.b AS test.b = Boolean(false)
  ...
```

```
Projection: Int32(0) AS CAST(Utf8(\"0\") AS Int32)
  ...
```


We have a lot tests that would be failed due to that.

Reply:
Have you tried following the model in https://github.com/apache/arrow-datafusion/pull/1315/files#diff-1d33be1a7e8231e53102eab8112e30aa89d8f5cb8c21cd25bcfbce3050cdb433R110 ? I think that calls `columnize_expr` among perhaps some other differences. 

Basically I think the code needs to do something like walk over the field names in the output `schema` and if they names of the rewritten exprs weren't the same add an alias;

Reply:
(I agree with @Dandandan  that this should apply to all nodes, not just a few special cased ones)

Reply:
> Basically I think the code needs to do something like walk over the field names in the output schema and if they names of the rewritten exprs weren't the same add an alias;

This sounds promising. No, I've not tried `columnize_expr`. Let me revise this and see if it works.

Reply:
I'm trying above approach. But must to note here is, it is still not for all nodes. For example, `Filter`'s schema is not related to its expressions. So we don't need (and shouldn't) add aliases for `Filter`.

Reply:
Not these approaches can work except the previous commit. Field name isn't equal to expression name for all cases. The cause of the problem is highly coupled with how projection push down rule works. Currently I don't find a solution to deal with all nodes other than current working version.

Reply:
Thanks for trying @viirya  -- I'll see if I can find some time this weekend to mess around with it

Reply:
Thanks @alamb . I'll keep trying on this too.

Reply:
> For example, for Project, it will create many (looks redundant) aliases. Some looks okay but some looks really weird, e.g. some failed tests:

@viirya the example you gave here looks like correct behavior to me, are you concerned with lots of updates on the tests? or are there other unwanted side effect of this approach?

Reply:
Oh, I'm simply unsure if such changes are okay here as it looks like most queries will be affected (not about its results but the cosmetic one).

Reply:
If it looks good for you, I will update all the tests.